File: Components/WinForms/CS/SchedulerMainDemo/Modules/EntityBoundMode.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Data.Entity;
using System.Configuration;
using System.Data.SQLite;
using System.Data.Common;
using System.ComponentModel.DataAnnotations;

namespace DevExpress.XtraScheduler.Demos {
    public partial class EntityBoundModeModule : TutorialControl {
        DoctorScheduleContext dataContext;
        public EntityBoundModeModule() {
            InitializeComponent();
            this.schedulerControl1.Start = DoctorScheduleDataHelper.BaseDate;
            Database.SetInitializer<DoctorScheduleContext>(null);
            this.dataContext = new DoctorScheduleContext();
            DoctorScheduleDataHelper.InitIfRequired(this.dataContext);

            InitializeMappings(this.schedulerStorage1);
            DoctorScheduleDataHelper.CustomizeLabelsAndStatuses(this.schedulerStorage1);

            this.dataContext.Doctors.Load();
            this.dataContext.DoctorAppointments.Load();

            this.schedulerStorage1.Resources.DataSource = this.dataContext.Doctors.Local.ToBindingList();
            this.schedulerStorage1.Appointments.DataSource = this.dataContext.DoctorAppointments.Local.ToBindingList();

            this.schedulerStorage1.AppointmentsChanged += schedulerStorage1_AppointmentsChanged;
            this.schedulerStorage1.AppointmentsInserted += schedulerStorage1_AppointmentsChanged;
            this.schedulerStorage1.AppointmentsDeleted += schedulerStorage1_AppointmentsChanged;
        }

        void schedulerStorage1_AppointmentsChanged(object sender, PersistentObjectsEventArgs e) {
            this.dataContext.SaveChanges();
        }
        void InitializeMappings(SchedulerStorage schedulerStorage) {
            AppointmentMappingInfo appointmentMapping = schedulerStorage.Appointments.Mappings;
            appointmentMapping.AllDay = "AllDay";
            appointmentMapping.Description = "Note";
            appointmentMapping.Subject = "PatientName";
            appointmentMapping.Start = "StartDate";
            appointmentMapping.End = "EndDate";
            appointmentMapping.Label = "IssueId";
            appointmentMapping.Location = "Location";
            appointmentMapping.RecurrenceInfo = "RecurrenceInfo";
            appointmentMapping.ReminderInfo = "ReminderInfo";
            appointmentMapping.ResourceId = "DoctorId";
            appointmentMapping.Status = "PaymentStatusId";
            appointmentMapping.Type = "EventType";

            ResourceMappingInfo resourceMapping = schedulerStorage.Resources.Mappings;
            resourceMapping.Id = "Id";
            resourceMapping.Caption = "Name";
        }
    }

    #region Data model
    public class DoctorAppointment {
        [Key]
        public Int64 Id { get; set; }
        public bool AllDay { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public string PatientName { get; set; }
        public string Note { get; set; }
        public int PaymentStatusId { get; set; }
        public int IssueId { get; set; }
        public int EventType { get; set; }
        public string Location { get; set; }
        public string RecurrenceInfo { get; set; }
        public string ReminderInfo { get; set; }
        public Int64? DoctorId { get; set; }
    }
    public class Doctor {
        [Key]
        public Int64 Id { get; set; }
        public string Name { get; set; }
    }

    public class DoctorScheduleContext : DbContext {
        public DoctorScheduleContext()
            : base("name=DevExpress.XtraScheduler.Demos.Properties.Settings.DoctorConnectionString") {

        }
        public DbSet<DoctorAppointment> DoctorAppointments { get; set; }
        public DbSet<Doctor> Doctors { get; set; }
    }
    #endregion

    #region DoctorScheduleDataHelper
    public class DoctorScheduleDataHelper {
        #region Table Creation Script
        public const string TableCreationScript =
     @"
-- Table: Doctors
CREATE TABLE Doctors ( 
Id      INTEGER PRIMARY KEY AUTOINCREMENT
                    UNIQUE,
    Name    TEXT
);

-- Table: DoctorAppointments
CREATE TABLE DoctorAppointments ( 
    Id             INTEGER  PRIMARY KEY AUTOINCREMENT
                            UNIQUE,
    AllDay          BOOLEAN,
    StartDate           DATETIME,
    EndDate           DATETIME,
    PatientName     TEXT,
    Note            TEXT,
    IssueId         INTEGER,
    PaymentStatusId INTEGER,
    EventType       INTEGER,
    Location        TEXT,
    RecurrenceInfo  TEXT,
    ReminderInfo    TEXT,
    DoctorId        INT64 
);
";
        #endregion
        public static DateTime BaseDate = DateTime.Today.AddDays(-1);
        public static string[] IssueList = { "Consultation", "Treatment", "X-Ray" };
        public static Color[] IssueColorList = { Color.SteelBlue, Color.Pink, Color.SeaShell };
        public static string[] PatientNames = { "Andrew Glover", "Mark Oliver", "Addison Davis", "Benjamin Hughes", "Lucas Smith" };
        public static string[] DoctorNames = { "Isabella Carter", "Miguel Simmons", "Madeleine Russell", "Ariana Alexander" };
        public static string[] PaymentStatuses = { "Paid", "Unpaid" };
        public static Color[] PaymentColorStatuses = { Color.Green, Color.Red };

        public static void InitIfRequired(DoctorScheduleContext dataContext) {
            DoctorScheduleDataHelper helper = new DoctorScheduleDataHelper(dataContext);
            helper.Create();
        }
        public static void CustomizeLabelsAndStatuses(SchedulerStorage storage) {
            AppointmentLabelCollection labels = storage.Appointments.Labels;
            labels.Clear();
            int count = IssueList.Length;
            for (int i = 0; i < count; i++) {
                IAppointmentLabel label = labels.CreateNewLabel(IssueList[i]);
                label.SetColor(IssueColorList[i]);
                labels.Add(label);
            }
            AppointmentStatusCollection statuses = storage.Appointments.Statuses;
            statuses.Clear();
            count = PaymentStatuses.Length;
            for (int i = 0; i < count; i++) {
                AppointmentStatus status = statuses.CreateNewStatus(PaymentStatuses[i]);
                status.Type = AppointmentStatusType.Custom;
                status.SetBrush(new SolidBrush(PaymentColorStatuses[i]));
                statuses.Add(status);
            }
        }

        Random rnd = new Random();
        DoctorScheduleContext dataContext;
        bool isSaveRequired = false;

        protected DoctorScheduleDataHelper(DoctorScheduleContext dataContext) {
            this.dataContext = dataContext;
        }

        public void Create() {
            CreateDBStructureIfRequired();
            this.dataContext.Doctors.Load();
            this.dataContext.DoctorAppointments.Load();
            if (this.dataContext.Doctors.Count() <= 0)
                CreateDoctors();

            if (this.dataContext.DoctorAppointments.Count() > 0) {
                DoctorAppointment appointment = this.dataContext.DoctorAppointments.FirstOrDefault((apt) => apt.EndDate >= DoctorScheduleDataHelper.BaseDate);
                if (appointment == null) {
                    this.dataContext.DoctorAppointments.Local.Clear();
                    this.dataContext.SaveChanges();
                }
            }

            if (this.dataContext.DoctorAppointments.Count() <= 0)
                CreateSchedule();

            if (!this.isSaveRequired)
                return;
            this.dataContext.SaveChanges();
            this.isSaveRequired = false;
        }
        void CreateDBStructureIfRequired() {
            SQLiteConnection connection = this.dataContext.Database.Connection as SQLiteConnection;
            if (connection == null)
                return;
            connection.Open();
            DataTable a = connection.GetSchema("Tables");
            if (a.Rows.Count > 0)
                return;
            SQLiteCommand command = new SQLiteCommand(DoctorScheduleDataHelper.TableCreationScript, connection);
            command.ExecuteNonQuery();
        }
        void CreateDoctors() {
            int doctorCount = DoctorNames.Length;
            var doctors = this.dataContext.Doctors;
            for (int i = 0; i < doctorCount; i++) {
                Doctor doctor = dataContext.Doctors.Create();
                doctor.Name = DoctorNames[i];
                doctors.Add(doctor);
            }
            this.isSaveRequired = true;
        }
        void CreateSchedule() {
            Random rnd = new Random();
            int doctorAppointmentsCount = dataContext.DoctorAppointments.Count();
            if (doctorAppointmentsCount > 0)
                return;
            int doctorsCount = DoctorNames.Length;
            for (int doctorId = 1; doctorId <= doctorsCount; doctorId++)
                CreateDoctorSchedule(doctorId);
            this.isSaveRequired = true;
        }
        void CreateDoctorSchedule(int doctorId) {
            for (DateTime start = BaseDate; start < BaseDate.AddDays(7); start += TimeSpan.FromDays(1)) {
                CreateDoctorAppointment(doctorId, start.AddHours(this.rnd.Next(9, 12)));
                CreateDoctorAppointment(doctorId, start.AddHours(this.rnd.Next(14, 16)));
                CreateDoctorAppointment(doctorId, start.AddHours(this.rnd.Next(18, 20)));
            }
        }
        void CreateDoctorAppointment(int doctorId, DateTime start) {
            DoctorAppointment doctorAppointment = dataContext.DoctorAppointments.Create();
            doctorAppointment.StartDate = start;
            doctorAppointment.EndDate = start.AddHours(this.rnd.Next(1, 3));
            doctorAppointment.IssueId = this.rnd.Next(0, 3);
            doctorAppointment.PaymentStatusId = this.rnd.Next(0, 2);
            int patientNameCount = PatientNames.Length;
            doctorAppointment.PatientName = PatientNames[this.rnd.Next(0, patientNameCount - 1)];
            doctorAppointment.DoctorId = doctorId;
            this.dataContext.DoctorAppointments.Add(doctorAppointment);
        }
    }
    #endregion
}

File: Components/WinForms/VB/SchedulerMainDemo/Modules/EntityBoundMode.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Data.Entity
Imports System.Configuration
Imports System.Data.SQLite
Imports System.Data.Common
Imports System.ComponentModel.DataAnnotations

Namespace DevExpress.XtraScheduler.Demos
	Partial Public Class EntityBoundModeModule
		Inherits TutorialControl
		Private dataContext As DoctorScheduleContext
		Public Sub New()
			InitializeComponent()
			Me.schedulerControl1.Start = DoctorScheduleDataHelper.BaseDate
			Database.SetInitializer(Of DoctorScheduleContext)(Nothing)
			Me.dataContext = New DoctorScheduleContext()
			DoctorScheduleDataHelper.InitIfRequired(Me.dataContext)

			InitializeMappings(Me.schedulerStorage1)
			DoctorScheduleDataHelper.CustomizeLabelsAndStatuses(Me.schedulerStorage1)

			Me.dataContext.Doctors.Load()
			Me.dataContext.DoctorAppointments.Load()

			Me.schedulerStorage1.Resources.DataSource = Me.dataContext.Doctors.Local.ToBindingList()
			Me.schedulerStorage1.Appointments.DataSource = Me.dataContext.DoctorAppointments.Local.ToBindingList()

			AddHandler Me.schedulerStorage1.AppointmentsChanged, AddressOf schedulerStorage1_AppointmentsChanged
			AddHandler Me.schedulerStorage1.AppointmentsInserted, AddressOf schedulerStorage1_AppointmentsChanged
			AddHandler Me.schedulerStorage1.AppointmentsDeleted, AddressOf schedulerStorage1_AppointmentsChanged
		End Sub

		Private Sub schedulerStorage1_AppointmentsChanged(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
			Me.dataContext.SaveChanges()
		End Sub
		Private Sub InitializeMappings(ByVal schedulerStorage As SchedulerStorage)
			Dim appointmentMapping As AppointmentMappingInfo = schedulerStorage.Appointments.Mappings
			appointmentMapping.AllDay = "AllDay"
			appointmentMapping.Description = "Note"
			appointmentMapping.Subject = "PatientName"
			appointmentMapping.Start = "StartDate"
			appointmentMapping.End = "EndDate"
			appointmentMapping.Label = "IssueId"
			appointmentMapping.Location = "Location"
			appointmentMapping.RecurrenceInfo = "RecurrenceInfo"
			appointmentMapping.ReminderInfo = "ReminderInfo"
			appointmentMapping.ResourceId = "DoctorId"
			appointmentMapping.Status = "PaymentStatusId"
			appointmentMapping.Type = "EventType"

			Dim resourceMapping As ResourceMappingInfo = schedulerStorage.Resources.Mappings
			resourceMapping.Id = "Id"
			resourceMapping.Caption = "Name"
		End Sub
	End Class

	#Region "Data model"
	Public Class DoctorAppointment
		<Key> _
		Public Property Id() As Int64
		Public Property AllDay() As Boolean
		Public Property StartDate() As DateTime
		Public Property EndDate() As DateTime
		Public Property PatientName() As String
		Public Property Note() As String
		Public Property PaymentStatusId() As Integer
		Public Property IssueId() As Integer
		Public Property EventType() As Integer
		Public Property Location() As String
		Public Property RecurrenceInfo() As String
		Public Property ReminderInfo() As String
		Private privateDoctorId? As Int64
		Public Property DoctorId() As Int64?
			Get
				Return privateDoctorId
			End Get
			Set(ByVal value? As Int64)
				privateDoctorId = value
			End Set
		End Property
	End Class
	Public Class Doctor
		<Key> _
		Public Property Id() As Int64
		Public Property Name() As String
	End Class

	Public Class DoctorScheduleContext
		Inherits DbContext
		Public Sub New()
			MyBase.New("name=DevExpress.XtraScheduler.Demos.Properties.Settings.DoctorConnectionString")

		End Sub
		Public Property DoctorAppointments() As DbSet(Of DoctorAppointment)
		Public Property Doctors() As DbSet(Of Doctor)
	End Class
	#End Region

	#Region "DoctorScheduleDataHelper"
	Public Class DoctorScheduleDataHelper
		#Region "Table Creation Script"
		Public Const TableCreationScript As String = "" & ControlChars.CrLf & "-- Table: Doctors" & ControlChars.CrLf & "CREATE TABLE Doctors ( " & ControlChars.CrLf & "Id      INTEGER PRIMARY KEY AUTOINCREMENT" & ControlChars.CrLf & "                    UNIQUE," & ControlChars.CrLf & "    Name    TEXT" & ControlChars.CrLf & ");" & ControlChars.CrLf & ControlChars.CrLf & "-- Table: DoctorAppointments" & ControlChars.CrLf & "CREATE TABLE DoctorAppointments ( " & ControlChars.CrLf & "    Id             INTEGER  PRIMARY KEY AUTOINCREMENT" & ControlChars.CrLf & "                            UNIQUE," & ControlChars.CrLf & "    AllDay          BOOLEAN," & ControlChars.CrLf & "    StartDate           DATETIME," & ControlChars.CrLf & "    EndDate           DATETIME," & ControlChars.CrLf & "    PatientName     TEXT," & ControlChars.CrLf & "    Note            TEXT," & ControlChars.CrLf & "    IssueId         INTEGER," & ControlChars.CrLf & "    PaymentStatusId INTEGER," & ControlChars.CrLf & "    EventType       INTEGER," & ControlChars.CrLf & "    Location        TEXT," & ControlChars.CrLf & "    RecurrenceInfo  TEXT," & ControlChars.CrLf & "    ReminderInfo    TEXT," & ControlChars.CrLf & "    DoctorId        INT64 " & ControlChars.CrLf & ");" & ControlChars.CrLf & ""
		#End Region
		Public Shared BaseDate As DateTime = DateTime.Today.AddDays(-1)
		Public Shared IssueList() As String = { "Consultation", "Treatment", "X-Ray" }
		Public Shared IssueColorList() As Color = { Color.SteelBlue, Color.Pink, Color.SeaShell }
		Public Shared PatientNames() As String = { "Andrew Glover", "Mark Oliver", "Addison Davis", "Benjamin Hughes", "Lucas Smith" }
		Public Shared DoctorNames() As String = { "Isabella Carter", "Miguel Simmons", "Madeleine Russell", "Ariana Alexander" }
		Public Shared PaymentStatuses() As String = { "Paid", "Unpaid" }
		Public Shared PaymentColorStatuses() As Color = { Color.Green, Color.Red }

		Public Shared Sub InitIfRequired(ByVal dataContext As DoctorScheduleContext)
			Dim helper As New DoctorScheduleDataHelper(dataContext)
			helper.Create()
		End Sub
		Public Shared Sub CustomizeLabelsAndStatuses(ByVal storage As SchedulerStorage)
			Dim labels As AppointmentLabelCollection = storage.Appointments.Labels
			labels.Clear()
			Dim count As Integer = IssueList.Length
			For i As Integer = 0 To count - 1
				Dim label As IAppointmentLabel = labels.CreateNewLabel(IssueList(i))
				label.SetColor(IssueColorList(i))
				labels.Add(label)
			Next i
			Dim statuses As AppointmentStatusCollection = storage.Appointments.Statuses
			statuses.Clear()
			count = PaymentStatuses.Length
			For i As Integer = 0 To count - 1
				Dim status As AppointmentStatus = statuses.CreateNewStatus(PaymentStatuses(i))
				status.Type = AppointmentStatusType.Custom
				status.SetBrush(New SolidBrush(PaymentColorStatuses(i)))
				statuses.Add(status)
			Next i
		End Sub

		Private rnd As New Random()
		Private dataContext As DoctorScheduleContext
		Private isSaveRequired As Boolean = False

		Protected Sub New(ByVal dataContext As DoctorScheduleContext)
			Me.dataContext = dataContext
		End Sub

		Public Sub Create()
			CreateDBStructureIfRequired()
			Me.dataContext.Doctors.Load()
			Me.dataContext.DoctorAppointments.Load()
			If Me.dataContext.Doctors.Count() <= 0 Then
				CreateDoctors()
			End If

			If Me.dataContext.DoctorAppointments.Count() > 0 Then
				Dim appointment As DoctorAppointment = Me.dataContext.DoctorAppointments.FirstOrDefault(Function(apt) apt.EndDate >= DoctorScheduleDataHelper.BaseDate)
				If appointment Is Nothing Then
					Me.dataContext.DoctorAppointments.Local.Clear()
					Me.dataContext.SaveChanges()
				End If
			End If

			If Me.dataContext.DoctorAppointments.Count() <= 0 Then
				CreateSchedule()
			End If

			If (Not Me.isSaveRequired) Then
				Return
			End If
			Me.dataContext.SaveChanges()
			Me.isSaveRequired = False
		End Sub
		Private Sub CreateDBStructureIfRequired()
			Dim connection As SQLiteConnection = TryCast(Me.dataContext.Database.Connection, SQLiteConnection)
			If connection Is Nothing Then
				Return
			End If
			connection.Open()
			Dim a As DataTable = connection.GetSchema("Tables")
			If a.Rows.Count > 0 Then
				Return
			End If
			Dim command As New SQLiteCommand(DoctorScheduleDataHelper.TableCreationScript, connection)
			command.ExecuteNonQuery()
		End Sub
		Private Sub CreateDoctors()
			Dim doctorCount As Integer = DoctorNames.Length
			Dim doctors = Me.dataContext.Doctors
			For i As Integer = 0 To doctorCount - 1
				Dim doctor As Doctor = dataContext.Doctors.Create()
				doctor.Name = DoctorNames(i)
				doctors.Add(doctor)
			Next i
			Me.isSaveRequired = True
		End Sub
		Private Sub CreateSchedule()
			Dim rnd As New Random()
			Dim doctorAppointmentsCount As Integer = dataContext.DoctorAppointments.Count()
			If doctorAppointmentsCount > 0 Then
				Return
			End If
			Dim doctorsCount As Integer = DoctorNames.Length
			For doctorId As Integer = 1 To doctorsCount
				CreateDoctorSchedule(doctorId)
			Next doctorId
			Me.isSaveRequired = True
		End Sub
		Private Sub CreateDoctorSchedule(ByVal doctorId As Integer)
			Dim start As DateTime = BaseDate
			Do While start < BaseDate.AddDays(7)
				CreateDoctorAppointment(doctorId, start.AddHours(Me.rnd.Next(9, 12)))
				CreateDoctorAppointment(doctorId, start.AddHours(Me.rnd.Next(14, 16)))
				CreateDoctorAppointment(doctorId, start.AddHours(Me.rnd.Next(18, 20)))
				start = start + TimeSpan.FromDays(1)
			Loop
		End Sub
		Private Sub CreateDoctorAppointment(ByVal doctorId As Integer, ByVal start As DateTime)
			Dim doctorAppointment As DoctorAppointment = dataContext.DoctorAppointments.Create()
			doctorAppointment.StartDate = start
			doctorAppointment.EndDate = start.AddHours(Me.rnd.Next(1, 3))
			doctorAppointment.IssueId = Me.rnd.Next(0, 3)
			doctorAppointment.PaymentStatusId = Me.rnd.Next(0, 2)
			Dim patientNameCount As Integer = PatientNames.Length
			doctorAppointment.PatientName = PatientNames(Me.rnd.Next(0, patientNameCount - 1))
			doctorAppointment.DoctorId = doctorId
			Me.dataContext.DoctorAppointments.Add(doctorAppointment)
		End Sub
	End Class
	#End Region
End Namespace
